Traceback (most recent call last):
  File "/Applications/anaconda3/envs/plumed-masterclass-2022/lib/python3.10/site-packages/jupyter_cache/executors/utils.py", line 51, in single_nb_execution
    executenb(
  File "/Applications/anaconda3/envs/plumed-masterclass-2022/lib/python3.10/site-packages/nbclient/client.py", line 1204, in execute
    return NotebookClient(nb=nb, resources=resources, km=km, **kwargs).execute()
  File "/Applications/anaconda3/envs/plumed-masterclass-2022/lib/python3.10/site-packages/nbclient/util.py", line 84, in wrapped
    return just_run(coro(*args, **kwargs))
  File "/Applications/anaconda3/envs/plumed-masterclass-2022/lib/python3.10/site-packages/nbclient/util.py", line 62, in just_run
    return loop.run_until_complete(coro)
  File "/Applications/anaconda3/envs/plumed-masterclass-2022/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/Applications/anaconda3/envs/plumed-masterclass-2022/lib/python3.10/site-packages/nbclient/client.py", line 663, in async_execute
    await self.async_execute_cell(
  File "/Applications/anaconda3/envs/plumed-masterclass-2022/lib/python3.10/site-packages/nbclient/client.py", line 965, in async_execute_cell
    await self._check_raise_for_error(cell, cell_index, exec_reply)
  File "/Applications/anaconda3/envs/plumed-masterclass-2022/lib/python3.10/site-packages/nbclient/client.py", line 862, in _check_raise_for_error
    raise CellExecutionError.from_cell_and_msg(cell, exec_reply_content)
nbclient.exceptions.CellExecutionError: An error occurred while executing the following cell:
------------------
## Solution trace
import scipy as sp


# Variables
A=15 * 1E4 # [m^2 * cm^2 / m^2]
P_r=5.5 * 101325 # [atm] * [Pa / atm]
P_p=1.01 * 101325 # [atm] * [Pa / atm] 
l=2.0*1E-4 # cm
rho_STP=1/22.4E3 # / [mol/cm3(stp)]
PCO2=0.034E-13 * 3600 # [cm3(STP) * cm/cm2/s/Pa] * [s/h]
PN2=0.089E-14 * 3600#  [cm3(STP) * cm/cm2/s/Pa] * [s/h]
# Permeate Composition
#y=0.43
# Feed Composition
z = 0.56

alpha=PCO2/PN2
Fin=0.02832

def equations(vars):
    x, y, Fp, Fr, theta = vars
    eq1 = ((theta - 1) / theta) * x + (z / theta) - y
    eq2 = x - (y * (1 + P_p/P_r * (1-y) * (alpha - 1)) / (alpha - (alpha - 1) * y))
    eq3 = Fin - Fr - Fp
    eq4 = Fp*y-A*PCO2*rho_STP / l * (x*P_r-y*P_p)
    eq5 = Fin - (Fp/theta)
    #eq4 = Fp*(1-y)-A*PN2*rho_STP / l * ((1-x)*P_r-(1-y)*P_p)
    return [eq1, eq2, eq3, eq4, eq5]

x, y, Fp, Fr, theta =  sp.optimize.fsolve(equations, (0.1, 0.1, 0.1, 0.1, 0.1))

print(equations([x, y, Fp, Fr, theta]))

print("\nFeed composition: ", f"{z:.4}", " [-]")
print("Permeate composition: ", f"{y:.4}", " [-]")
print("Retentate composition: ", f"{x:.4}", " [-]")
print("Cut: ", f"{theta:.4}", "[-]")
print("Permeate flowrate: ", f"{Fp:.4}", " [mol/h]")
print("Feed flowrate: ", f"{Fin:.4}",  " [mol/h]")
print("Retentate flowrate: ", f"{Fr:.4}", " [mol/h]\n")
------------------

[0;31m---------------------------------------------------------------------------[0m
[0;31mAttributeError[0m                            Traceback (most recent call last)
Input [0;32mIn [1][0m, in [0;36m<cell line: 31>[0;34m()[0m
[1;32m     28[0m     [38;5;66;03m#eq4 = Fp*(1-y)-A*PN2*rho_STP / l * ((1-x)*P_r-(1-y)*P_p)[39;00m
[1;32m     29[0m     [38;5;28;01mreturn[39;00m [eq1, eq2, eq3, eq4, eq5]
[0;32m---> 31[0m x, y, Fp, Fr, theta [38;5;241m=[39m  [43msp[49m[38;5;241;43m.[39;49m[43moptimize[49m[38;5;241m.[39mfsolve(equations, ([38;5;241m0.1[39m, [38;5;241m0.1[39m, [38;5;241m0.1[39m, [38;5;241m0.1[39m, [38;5;241m0.1[39m))
[1;32m     33[0m [38;5;28mprint[39m(equations([x, y, Fp, Fr, theta]))
[1;32m     35[0m [38;5;28mprint[39m([38;5;124m"[39m[38;5;130;01m\n[39;00m[38;5;124mFeed composition: [39m[38;5;124m"[39m, [38;5;124mf[39m[38;5;124m"[39m[38;5;132;01m{[39;00mz[38;5;132;01m:[39;00m[38;5;124m.4[39m[38;5;132;01m}[39;00m[38;5;124m"[39m, [38;5;124m"[39m[38;5;124m [-][39m[38;5;124m"[39m)

[0;31mAttributeError[0m: module 'scipy' has no attribute 'optimize'
AttributeError: module 'scipy' has no attribute 'optimize'

